diff --git a/cosmic-applet-audio/src/lib.rs b/cosmic-applet-audio/src/lib.rs
index 5f49d0f9..0e698013 100644
--- a/cosmic-applet-audio/src/lib.rs
+++ b/cosmic-applet-audio/src/lib.rs
@@ -38,8 +38,22 @@ const GO_BACK: &str = "media-skip-backward-symbolic";
 const GO_NEXT: &str = "media-skip-forward-symbolic";
 const PAUSE: &str = "media-playback-pause-symbolic";
 const PLAY: &str = "media-playback-start-symbolic";
+const VERSION: &str = env!("CARGO_PKG_VERSION");
+// use std::fs::File;
 
 pub fn run() -> cosmic::iced::Result {
+    // let _ = tracing_log::LogTracer::init();
+
+    // let _ = tracing_subscriber::fmt()
+    //     .with_max_level(tracing::Level::TRACE)
+    //     .try_init();
+
+    tracing::info!("=== INFO INFO2 Starting audio applet with version {VERSION} ===");
+    tracing::debug!("=== DEBUG DEBUG2!!!! Starting audio applet with version {VERSION} ===");
+
+    eprintln!("=== STDERR TEST2 ===");
+    println!("=== STDOUT2 TEST ===");
+    tracing::error!("=== TRACING ERROR2 ===");
     localize();
     cosmic::applet::run::<Audio>(())
 }
@@ -325,6 +339,10 @@ impl cosmic::Application for Audio {
             }
 
             Message::SetDefaultSink(pos) => {
+                    eprintln!("=== Message eprintl::SetDefaultSink ===");
+                    println!("=== Message::printl SetDefaultSink ===");
+                    tracing::error!("=== tracing Message::SetDefaultSink ===");
+
                 return self
                     .model
                     .set_default_sink(pos)
diff --git a/cosmic-applets/src/main.rs b/cosmic-applets/src/main.rs
index f62bbf21..84855a52 100644
--- a/cosmic-applets/src/main.rs
+++ b/cosmic-applets/src/main.rs
@@ -4,7 +4,10 @@
 const VERSION: &str = env!("CARGO_PKG_VERSION");
 
 fn main() -> cosmic::iced::Result {
-    tracing_subscriber::fmt().with_env_filter("warn").init();
+    // tracing_subscriber::fmt().with_env_filter("info").init();
+    tracing_subscriber::fmt().with_env_filter
+        (tracing_subscriber::EnvFilter::from_default_env()
+            ).init();
     let _ = tracing_log::LogTracer::init();
 
     let Some(applet) = std::env::args().next() else {
