diff --git a/hw/misc/ivshmem.c b/hw/misc/ivshmem.c
index d66d912172..ffcb5a18d0 100644
--- a/hw/misc/ivshmem.c
+++ b/hw/misc/ivshmem.c
@@ -498,6 +498,8 @@ static void process_msg_shmem(IVShmemState *s, int fd, Error **errp)
     /* mmap the region and map into the BAR2 */
     memory_region_init_ram_from_fd(&s->server_bar2, OBJECT(s), "ivshmem.bar2",
                                    size, RAM_SHARED, fd, 0, &local_err);
+    memory_region_add_subregion(get_system_memory(), (hwaddr) 0x1000000000, &s->server_bar2);
+
     if (local_err) {
         error_propagate(errp, local_err);
         return;